// @web3modal/ui/dist/esm/src/composites/wui-list-transaction/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, LitElement } from 'lit';
import { property } from 'lit/decorators.js';
import '../../components/wui-text/index.js';
import { elementStyles, resetStyles } from '../../utils/ThemeUtil.js';
import { UiHelperUtil } from '../../utils/UiHelperUtil.js';
import { customElement } from '../../utils/WebComponentsUtil.js';
import '../wui-transaction-visual/index.js';
import styles from './styles.js';
let WuiListTransaction = class WuiListTransaction extends LitElement {
    constructor() {
        super(...arguments);
        this.type = 'bought';
        this.disabled = false;
        this.imageSrc = '';
        this.date = new Date();
        this.transactionDetail = '';
    }
    render() {
        const isSent = this.type === 'nftSent' || this.type === 'cryptoSent';
        const title = isSent ? 'Sent' : this.type;
        const formattedDate = UiHelperUtil.getFormattedDate(this.date);
        return html `
      <button ?disabled=${this.disabled} ontouchstart>
        <wui-transaction-visual
          type=${this.type}
          imageSrc=${this.imageSrc}
        ></wui-transaction-visual>
        <wui-flex flexDirection="column" gap="3xs">
          <wui-text variant="paragraph-600" color="fg-100">${title}</wui-text>
          <wui-text variant="small-500" color="fg-200">${this.transactionDetail}</wui-text>
        </wui-flex>
        <wui-text variant="micro-700" color="fg-300">${formattedDate}</wui-text>
      </button>
    `;
    }
};
WuiListTransaction.styles = [resetStyles, elementStyles, styles];
__decorate([
    property()
], WuiListTransaction.prototype, "type", void 0);
__decorate([
    property({ type: Boolean })
], WuiListTransaction.prototype, "disabled", void 0);
__decorate([
    property()
], WuiListTransaction.prototype, "imageSrc", void 0);
__decorate([
    property({ attribute: false })
], WuiListTransaction.prototype, "date", void 0);
__decorate([
    property()
], WuiListTransaction.prototype, "transactionDetail", void 0);
WuiListTransaction = __decorate([
    customElement('wui-list-transaction')
], WuiListTransaction);
export { WuiListTransaction };
