// @web3modal/wagmi/dist/esm/exports/vue.js
import { onUnmounted, reactive, ref } from 'vue';
import { Web3Modal } from '../src/client.js';
import { VERSION } from '../src/utils/constants.js';
let modal = undefined;
export function createWeb3Modal(options) {
    if (!modal) {
        modal = new Web3Modal({ ...options, _sdkVersion: `vue-wagmi-${VERSION}` });
    }
    return modal;
}
export function useWeb3ModalTheme() {
    if (!modal) {
        throw new Error('Please call "createWeb3Modal" before using "useWeb3ModalTheme" hook');
    }
    function setThemeMode(themeMode) {
        modal?.setThemeMode(themeMode);
    }
    function setThemeVariables(themeVariables) {
        modal?.setThemeVariables(themeVariables);
    }
    const themeVariables = ref(modal.getThemeVariables());
    const themeMode = ref(modal.getThemeMode());
    const unsubscribe = modal?.subscribeTheme(state => {
        themeMode.value = state.themeMode;
        themeVariables.value = state.themeVariables;
    });
    onUnmounted(() => {
        unsubscribe?.();
    });
    return {
        setThemeMode,
        setThemeVariables,
        themeMode,
        themeVariables
    };
}
export function useWeb3Modal() {
    if (!modal) {
        throw new Error('Please call "createWeb3Modal" before using "useWeb3Modal" composable');
    }
    async function open(options) {
        await modal?.open(options);
    }
    async function close() {
        await modal?.close();
    }
    return reactive({
        open,
        close
    });
}
export function useWeb3ModalState() {
    if (!modal) {
        throw new Error('Please call "createWeb3Modal" before using "useWeb3ModalState" composable');
    }
    const state = reactive(modal.getState());
    const unsubscribe = modal?.subscribeState(next => {
        state.open = next.open;
        state.selectedNetworkId = next.selectedNetworkId;
    });
    onUnmounted(() => {
        unsubscribe?.();
    });
    return state;
}
export function useWeb3ModalEvents() {
    if (!modal) {
        throw new Error('Please call "createWeb3Modal" before using "useWeb3ModalEvents" composable');
    }
    const event = reactive(modal.getEvent());
    const unsubscribe = modal?.subscribeEvents(next => {
        event.data = next.data;
        event.timestamp = next.timestamp;
    });
    onUnmounted(() => {
        unsubscribe?.();
    });
    return event;
}
export { EIP6963Connector } from '../src/connectors/EIP6963Connector.js';
export { defaultWagmiConfig } from '../src/utils/defaultWagmiCoreConfig.js';
